Lib/functools.py

```diff
@@ -274,43 +275,129 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
-# Purely functional, no descriptor behaviour
-class partial:
-    """New function with partial application of the given arguments
-    and keywords.
+
+class _PlaceholderType:
+    """The type of the Placeholder singleton.
+
+    Used as a placeholder for partial arguments.
     """
+    __instance = None
+    __slots__ = ()
 
-    __slots__ = "func", "args", "keywords", "__dict__", "__weakref__"
+    def __init_subclass__(cls, *args, **kwargs):
+        raise TypeError(f"type '{cls.__name__}' is not an acceptable base type")
 
-    def __new__(cls, func, /, *args, **keywords):
+    def __new__(cls):
+        if cls.__instance is None:
+            cls.__instance = object.__new__(cls)
+        return cls.__instance
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __reduce__(self):
+        return 'Placeholder'
+
+Placeholder = _PlaceholderType()
+
+def _partial_prepare_merger(args):
+    nargs = len(args)
+    if not nargs:
+        return 0, None
+    order = list()
+    i, j = 0, nargs
+    for a in args:
```

Comment:
You can use `enumerate()` here.

```suggestion
    for i, a in enumerate(args):
```